File: week1/maxPairWise.js
const { performance } = require("perf_hooks");

// max pairwise product

//nested loops
function naiveAlgorithm(arr) {
  let product = 0;
  for (el of arr) {
    for (i of arr) {
      let val = el * i;
      let test = el !== i && val > product;
      if (test) {
        product = val;
      }
    }
  }
  return product;
}

// two separate loops to find largest and next largest numbers
function maxPairFast(arr) {
  let largest = arr[0];
  let largestInx = 0;
  let large = arr[0];
  arr.forEach((el, inx) => {
    if (el > largest) {
      largest = el;
      largestInx = inx;
    }
  });
  arr.forEach((el, inx) => {
    if (el > large && inx !== largestInx) {
      large = el;
    }
  });
  return large * largest;
}

//reduce method
function maxPairReduce(arr) {
  const data = arr.reduce(
    (acc, cur, inx, arry) => {
      if (cur > acc.largest) {
        acc.largestInx = inx;
        acc.nextLargest = acc.largest;
        acc.largest = cur;
      } else if (cur > acc.nextLargest && cur !== acc.largest) {
        acc.nextLargest = cur;
      }
      return acc;
    },
    { largest: 0, largestInx: 0, nextLargest: 0 }
  );
  return data.largest * data.nextLargest;
}

// returns the amount of time a function takes

function timer(func, testData) {
  const start = performance.now();
  func(testData);
  const end = performance.now();
  return end - start;
}

function logPerformace(funcArr, testData) {
  funcArr.forEach(obj => {
    const name = Object.keys(obj)[0];
    const func = obj[name];
    console.log(`${name} returns in ${timer(func, testData)} `);
  });
}

const funcArr = [
  { maxPairFast: maxPairFast },
  { maxPairReduce: maxPairReduce },
  { naiveAlgorithm: naiveAlgorithm }
];

function generateTest(numberOfTests, maxArrLength) {
  //Each iteration is a test
  for (let i = 0; i < numberOfTests; i++) {
    let testArr = [];
    let length = Math.floor(Math.random() * Math.floor(maxArrLength));
    console.log(`testing functions with an array length ${length}`);
    //build testArr
    for (let j = 0; j < length; j++) {
      testArr.push(Math.floor(Math.random() * Math.floor(5)));
    }
    //run test
    logPerformace(funcArr, testArr);
  }
}
console.log(
  "testing functions with 10 randomly generated arrays with max length 100"
);
generateTest(10, 100);

File: week1/maxPairWise.py
import time

n = input("array: ")


def naive(arr):
    result = 0
    for el in arr:
        for i in arr:
            if arr.index(el) != arr.index(i) and (i * el) > result:
                result = el * i
    return result


def faster(arr):
    largest = arr[0]
    large = arr[0]
    for el in arr:
        if el > largest:
            largest = el
    for foo in arr:
        if foo > large and arr.index(foo) != arr.index(largest):
            large = foo
    return large * largest


print(naive(n))
print(faster(n))


def timer(func, arr):
    start = time.time()
    func(arr)
    end = time.time()
    result = end - start
    return format(result, "f")


testArr = [111, 44, 33, 78, 9, 2, 2345, 21, 5, 3, 1, 54, 987, 5555, 43, 56565]

print("naive algorithm timed at ", timer(naive, testArr))

print("faster algorithm timed at ", timer(faster, testArr))
